torch/csrc/jit/mobile/backport_factory.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+//#include <ATen/core/ivalue.h>
+#include <functional>
+#include <memory>
+#include <unordered_map>
+
+namespace c10 {
+struct IValue;
+}
+
+namespace caffe2 {
+namespace serialize {
+class IStreamAdapter;
+class ReadAdapterInterface;
+class PyTorchStreamWriter;
+class PyTorchStreamReader;
+} // namespace serialize
+} // namespace caffe2
+
+namespace torch {
+namespace jit {
+
+constexpr const char* kArchiveNameConstants = "constants";
+constexpr const char* kArchiveNameBytecode = "bytecode";
+constexpr const char* kArchiveNameVersion = "version";
+constexpr int64_t kBytecodeVersionV4 = 0x4L;
+constexpr int64_t kBytecodeVersionV5 = 0x5L;
+
+/*
+BackportFactory manages a list of backport from n to n-1 function, and provides
+function to check if a specific function exists.
+*/
+class BackportFactory final {
```

Comment:
I'm finding the current code organization a bit confusing:

* There's BackportFactory which basically serves as the entry point for the backport functionality across all versions. Not sure a class was needed per se, but if the goal is to aggregate all backport functions (because all registrations will happen within in it) and serve as the only implementation of the backport logic then I think it should be renamed (e.g. BackportManager).
* The backport_factory.cpp does register the backport functions but the implementations live in a separate file (e.g. backport_v5_to_v4.cpp), which is linked "on the side" (i.e. there's no direct dependency indicating that you can find the implementations needed for BackendFactory to work in those other files). I find that hard to follow.
* There's also those utility functions that get defined also in backport_factory.h but implemented in another .cpp (backport_common.cpp).

I see 2 better options:

**Option 1: Consolidate everything in BackendFactory>BackportManager**
If the goal is to have a single class that knows everything, then BackportManager should take all the implementations of backport functions.
This way the utility functions can be just defined in an anonymous namespace in the .cpp implementation.

So, backport_manager.cpp should contain:
* utilities under anonymous namespace.
* the backport registry implementation functions
* one backport from-to function implementation for each version bump
* Nothing more! If specific backport function implementations need to define more utilities then they should go in the anonymous namespace and ideally also be actual utilities (useful to other backport functions) --otherwise they should be lambdas or something inside the function implementation :-)

The reasoning behind doing this is based on the following assumptions:
* There's not going to be many backport functions.
* There's always only 1 of each (i.e. only 1 implementation of backport v5_to_v4).
* It would be simpler to manage having this in a single place (people can easy see how previous backports were added, there's only a single place to track for this functionality).  
* Only us in PyTorch will add backport functions (it's not like PyTorch users will have their own implementations).

Pros:
* There's a single dependency (backport_manager.h/.cpp) needed for everything!

Cons:
* The .cpp can get big.


**Option 2: Separate the BackportManager, the implementations and the utilities.**
BackportManager only functions as the agregator, and the implementation of the generic backport(from, to) function.
Backport functions definition, implementation and registration live in their own separate files; or perhaps better functions are implemented just taken as dependencies from BackportManager and registration still happens there, but at least the dependency to the specific functions is explicit.
The utilities are defined and implemented in their own files (no relation to BackportManager). Individual backport function implementations are free to use them or not.


Pros:
* Cleaner and clearer dependencies. More flexible dependency management.
* A separate space for each backport function implementation.

Cons:
* Many more files, and more things to add/modify when a new function is added.
* Arguably unnecessary since backports are stricktly defined one per version in the bytecode, which we manage internally. 


I am more inclined for Options 1 (we can always refactor to split them if it starts getting busy).
What do others think?
@iseeyuan @kimishpatel

Reply:
Option 2 was my intention, but yeah I agree the linkage isn't clear at this moment. I wasn't sure what will be the best way to register the function. 

Option 1 is clear to me. @iseeyuan brought up a point that, with better dependency, it's easier to clean the file when we need to retire one version.

Reply:
I fully agree the three points Raziel pointed out: 
* renaming to BackportManager
* implementation and registration in the same file. Please refer to https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/test_backend_compiler_preprocess.cpp
* .h and .cpp correspondence.

For the two options I was also thinking of option 2, referring to how we defined the preprocess registration. However, as Razeil mentioned above, backport is more internal, comparing to the public API of preprocess, which can be registered outside of PyTorch. Putting them in a single file has advantage that we only update codes not files - and it makes sense because it's internal and we don't have many backports. 

I vote option 1. Later if the file grows it's a simple refactor to split it.

Reply:
I am reading this and will comment soon. Have a vaccine appointment so maybe delayed a bit but wanted to update here.

Reply:
Agree with all the points made by @raziel. 

However of the two options my preference is the second one. 
Reasons:
- As @iseeyuan mentioned, it is easier to cleanup when we retire some version. Probably not that big of a deal though.
- More readable and functionalities are appropriately split. If we dont want to manage too many `backport_n_to_n-1.cpp`, we can have 1) backport_impl.cpp and 2) backport_manager.cpp. This way backport_manager is purely orchestrates backporting without having to be aware of the implementation. Downside of this is that, only consume of bcakport_impl is backport_manager, but we can put backport_impl under namespace like torch/mobile/backport/impl.

Another pattern I would encourage, and I think this was mentioned already, is to have registration macros like:
`REGISTER_BACKPORT_FUNCTION(model_version, function)` and this happens in backport_impl.cpp.
Furthermore for backporting from  from_version to to_version I would use a free function `backport(from_version, to_version, model)` and everything about backport manager is hidden inside this. This function itself lives in backport_manager.cpp. This way callsites dont have to be aware of backport manager.

Reply:
I can also go for option 1) for now and have a follow up to implement option 2) if that's ok.